Makefile

```diff
@@ -1,59 +1,149 @@
+# Project variables
 NAME        := terraform-docs
 VENDOR      := segmentio
 DESCRIPTION := Generate docs from Terraform modules
 MAINTAINER  := Martin Etmajer <metmajer@getcloudnative.io>
 URL         := https://github.com/$(VENDOR)/$(NAME)
 LICENSE     := MIT
 
-VERSION     := $(shell cat ./VERSION)
+# Repository variables
+PACKAGE     := github.com/$(VENDOR)/$(NAME)
 
-GOBUILD     := go build -ldflags "-X main.version=$(VERSION)"
-GOPKGS      := $(shell go list ./... | grep -v /vendor)
+# Build variables
+BUILD_DIR   := bin
+COMMIT_HASH ?= $(shell git rev-parse --short HEAD 2>/dev/null)
+VERSION     ?= $(shell git describe --tags --exact-match 2>/dev/null || git describe --tags 2>/dev/null || echo "v0.0.0-$(COMMIT_HASH)")
+BUILD_DATE  ?= $(shell date +%FT%T%z)
 
+# Go variables
+GOOS        ?= $(shell go env GOOS)
+GOARCH      ?= $(shell go env GOARCH)
+GOCMD       := GO111MODULE=on go
+MODVENDOR   := -mod=vendor
+GOFILES     ?= $(shell find . -type f -name '*.go' -not -path "./vendor/*")
+GOPKGS      ?= $(shell $(GOCMD) list $(MODVENDOR) ./... | grep -v /vendor)
+
+GOLDFLAGS   :="
+GOLDFLAGS   += -X $(PACKAGE)/internal/pkg/version.version=$(VERSION)
+GOLDFLAGS   += -X $(PACKAGE)/internal/pkg/version.commitHash=$(COMMIT_HASH)
+GOLDFLAGS   += -X $(PACKAGE)/internal/pkg/version.buildDate=$(BUILD_DATE)
+GOLDFLAGS   +="
+
+GOBUILD     ?= CGO_ENABLED=0 $(GOCMD) build $(MODVENDOR) -ldflags $(GOLDFLAGS)
+GORUN       ?= GOOS=$(GOOS) GOARCH=$(GOARCH) $(GOCMD) run $(MODVENDOR)
+
+# Binary versions
+GOLANGCI_VERSION  := v1.17.1
+GITCHGLOG_VERSION := 0.8.0
 
 .PHONY: all
 all: clean deps lint test build
 
-.PHONY: authors
-authors:
-	git log --all --format='%aN <%aE>' | sort -u | egrep -v noreply > AUTHORS
-
-.PHONY: build
-build: authors build-darwin-amd64 build-freebsd-amd64 build-linux-amd64 build-windows-amd64
+#########################
+## Development targets ##
+#########################
+.PHONY: clean
+clean: ## Clean workspace
+	@ $(MAKE) --no-print-directory log-$@
+	rm -rf ./$(BUILD_DIR)
 
-build-darwin-amd64:
-	GOOS=darwin GOARCH=amd64 $(GOBUILD) -o bin/darwin-amd64/$(NAME)
+.PHONY: vendor
+vendor: ## Install 'vendor' dependencies
+	@ $(MAKE) --no-print-directory log-$@
+	$(GOCMD) mod vendor
 
-build-freebsd-amd64:
-	GOOS=freebsd GOARCH=amd64 $(GOBUILD) -o bin/freebsd-amd64/$(NAME)
+.PHONY: verify
+verify: ## Verify 'vendor' dependencies
+	@ $(MAKE) --no-print-directory log-$@
+	$(GOCMD) mod verify
 
-build-linux-amd64:
-	GOOS=linux GOARCH=amd64 $(GOBUILD) -o bin/linux-amd64/$(NAME)
+.PHONY: lint
```

Comment:
Please sort the make targets in each block.

Reply:
+1